Client/TxPool: add tx validation

jochem-brouwer opened the pull request:
Closes #1850

jochem-brouwer commented on packages/client/lib/sync/txpool.ts:
```diff
@@ -165,16 +177,38 @@ export class TxPool {
    * @param tx Transaction
    */
-  add(tx: TypedTransaction) {
+  async add(tx: TypedTransaction) {
```
@ryanio do you think this method is a good central place for most checks? We can handle the throws internally here if we have a list of txs (remove them from list since these are invalid), but for RPC we catch those and instead throw the RPC error

ryanio replied:
yes, but i'm thinking make a `validate(tx: TypedTransaction)` method in this class, and call into it at the top of add.

jochem-brouwer replied:
Yep, makes sense, will do :+1: :smile:

ryanio commented on packages/client/lib/sync/txpool.ts:
```diff
@@ -119,6 +124,11 @@ export class TxPool {
+  /**
+   * 
+   */
+  private minGasPriceBump = 10 // percentage
```
nice yeah i don't think this needs to be a configurable opt, just a constant, like MIN_GAS_PRICE_BUMP_PERCENT

ryanio commented on packages/client/lib/rpc/modules/eth.ts:
```diff
@@ -931,7 +931,8 @@ export class Eth {
     // Add the tx to own tx pool
     const { txPool } = this.service.synchronizer as FullSynchronizer
-    txPool.add(tx)
+    // TODO handle any errors here
+    await txPool.add(tx)
```
yep i think this just can be

```suggestion
    try {
      await txPool.add(tx)
    } catch (error: any) {
      throw {
        code: INVALID_PARAMS
        message: error?.message ?? error.toString()
      }
    }
```

ryanio commented on packages/client/lib/sync/txpool.ts:
```diff
@@ -165,16 +177,38 @@ export class TxPool {
-      add = inPool.filter((poolObj) => !poolObj.tx.nonce.eq(tx.nonce))
+      let existingTxn = inPool.find((poolObj) => poolObj.tx.nonce.eq(tx.nonce))
+      if (existingTxn) {
+        let currentGasPrice
+        switch (existingTxn.tx.type) {
+          case 0: 
+            currentGasPrice = (existingTxn.tx as Transaction).gasPrice
+            break
+          case 1: 
+            currentGasPrice = (existingTxn.tx as AccessListEIP2930Transaction).gasPrice
+            break
+          case 2:
+            currentGasPrice = (existingTxn.tx as FeeMarketEIP1559Transaction).maxFeePerGas
+            break
+          default: 
+            throw new Error("unknown tx type") // TODO make this error better
+        }
+        let minimumGasPrice = currentGasPrice.add(currentGasPrice.muln(this.minGasPriceBump).divn(100))
```
hm that's a lot of boilerplate code just to get the tx gasPrice. would be nice to have a tx helper for this in the base tx class, can use the tx.supports(Capability) API to return the right response? (basically maxFeePerGas if FeeMarket, else gasPrice)

jochem-brouwer replied:
I had exactly the same idea, I just took a look how we do this in VM, we could refactor [there](https://github.com/ethereumjs/ethereumjs-monorepo/blob/07a6d2e8a672d6bebad57a1e9df80c3e735fbf2e/packages/vm/src/runTx.ts#L356) (this gasPrice propagates via `TxContext` to `EVM` to `EEI`, where it is used to get the gas price in the `GASPRICE` opcode) as well, the only thing is that the helper needs a `baseFee` parameter in case of a 1559 txn.

Something like `getGasPrice(baseFee?: BN)` on `BaseTransaction`, on 1559 we know that we are in a 1559 block which thus has `baseFee` and otherwise in 2930/Legacy txs we can just make it `undefined` since it points to `gasPrice`. (This whole `baseFee` stuff also makes TxPool more interesting since you could include a tx in the pool which could be mined now, but at some later blocks not if `baseFee` rises above the `maxPriorityFeePerGas` ... :thinking: )

ryanio replied:
Yeah that would be great!

The baseFee does add an interesting dynamic, so I don't think we should reject txs if they are below the baseFee, since it could eventually lower, but it would be nice on the rpc layer to know if a tx I am submitting is severely underpriced (meaning, it has very little chance to be mined). To start we can probably just model around how geth is already thinking about it, rather than coming up with our own rational, to best match their tx pool credentials.

ryanio commented on packages/client/lib/sync/txpool.ts:
```diff
@@ -325,7 +359,7 @@ export class TxPool {
     const newTxHashes = []
     for (const tx of txs) {
-      this.add(tx)
+      await this.add(tx)
```
here (and below on L418) you can add try catch to `await this.add(tx)`, and in the catch just do a ```logger.info(`Error adding tx to TxPool: ${error.message}`)``` and continue, not much else we need to do if we get a bad tx from devp2p. we could do a `logger.error`, but i think that is too verbose, for something that is out of our control.

jochem-brouwer commented:
Merged latest `master` changes so also have the TxPool extraction in it. 

Most of the validation logic is now in there, this of course has to be tested. Tests are now also broken due to these stricter validation rules.

jochem-brouwer commented:
TODOs:

- [ ] Ensure that the added tests check if adding invalid txs to TxPool throws, but also that this throws due to the right reason (check error message)
- [ ] Move the gas price calculation to Tx (however, `FeeMarketEIP1559Transaction` needs both `maxPriorityFeePerGas` and `maxFeePerGas` bumped by 10% in case user wants to override the txn)
- [ ] Fix bug in `@ethereumjs/tx` where tx options do not propagate from unsiged tx to signed tx.

jochem-brouwer commented:
I added the fee bump checks for all txs. I had to look into Geth how they do this. They have getters for each tx type which returns the "tip" and the "max fee" for each tx (so for Legacy and AccessLists it just returns `gasPrice`). I have implemented that here as well.

See for instance https://github.com/ethereum/go-ethereum/blob/af806168b62b6a0e772b73ce25d31d8e61d26043/core/types/legacy_tx.go#L101

Also added a check to see if tx can pay current basefee.

jochem-brouwer commented:
I think this is ready for a new round of reviews. However, I somewhat expect that the coverage is not 100% so might want to add more tests (will therefore keep this PR as draft PR)

ryanio commented on packages/client/lib/service/txpool.ts:
```diff
@@ -398,7 +540,7 @@ export class TxPool {
     const newTxHashes = []
     for (const tx of txs) {
-      this.add(tx)
+      await this.add(tx)
```
this needs a try/catch around it in case the tx is invalid, we don't want to abort processing the other txs

jochem-brouwer replied:
Ouch, good catch :open_mouth: :+1:

ryanio commented on packages/client/lib/service/txpool.ts:
```diff
@@ -1,18 +1,25 @@
+const MIN_GAS_PRICE_BUMP_PERCENT = 10
+const TX_MAX_DATA_SIZE = 128 * 1024 // 128KB
+const MAX_POOL_SIZE = 5000
+const MIN_GASPRICE = new BN(1000000000) // 1 GWei
```
nit, the first constant is `MIN_GAS_PRICE_BUMP_PERCENT` so this one should also have a separator between GAS_PRICE so `MIN_GAS_PRICE` (can also bring it up to the top of the list)

ryanio commented on packages/client/lib/service/txpool.ts:
```diff
@@ -73,6 +80,12 @@ export class TxPool {
+  /**
+   * The number of txs currently in the pool
+   */
+
```
extra newline
```suggestion
```

ryanio commented on packages/client/lib/service/txpool.ts:
```diff
@@ -178,27 +192,148 @@ export class TxPool {
+      }
+      // Local txs are not checked against MIN_GASPRICE
+      if (currentGasPrice.lt(MIN_GASPRICE)) {
+        throw new Error(`Tx does not pay the minimum gas price of ${MIN_GASPRICE.toString()}`)
```
```suggestion
        throw new Error(`Tx does not pay the minimum gas price of ${MIN_GASPRICE}`)
```
you don't need to call `toString()` in template literals it will automatically call it for you. it would be great if you can also remove in the other places in this file.

ryanio commented on packages/client/lib/service/txpool.ts:
```diff
@@ -178,27 +192,148 @@ export class TxPool {
    * If there is a tx in the pool with the same address and
-   * nonce it will be replaced by the new tx.
+   * nonce it will be replaced by the new tx, if it has a sufficent gas bump.
+   * (TODO) verifies certain constraints, if these are not matches, throws.
```
todo comment still here

ryanio commented:
looking great! just a few last comments from me, and maybe a last look for some additional coverage/test cases, then i'd say this is pretty ready! thanks for working on it, really cool stuff, big step in sophistication for our tx pool :)

jochem-brouwer commented on packages/client/lib/service/txpool.ts:
```diff
@@ -178,27 +192,148 @@ export class TxPool {
+  /**
+   * Returns the gas tip (the max priority fee) for each transaction type
+   * @param tx - tx to check
+   * @returns - max priority fee
```
I keep forgetting this, thanks :smile:

jochem-brouwer commented:
Addressed the review items, thanks! Code coverage should now cover all lines (except the log line but I don't think that is important).

ryanio commented:
looks excellent!

ryanio commented on packages/client/test/miner/miner.spec.ts:
```diff
@@ -325,11 +329,25 @@ tape('[Miner]', async (t) => {
-      txPool.add(createTx())
+      // In order not to pollute TxPool with too many txs from the same address
+      // (or txs which are already known), keep generating a new address for each tx
+      const address = new Address(privateToAddress(pkey))
```
btw we have this method
```suggestion
      const address = Address.fromPrivateKey(pkey)
```

ryanio reviewed (approved):
lgtm! great work
